platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt

```diff
@@ -142,149 +169,142 @@ sealed class PropertyTrace {
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) {
+            with(builder) {
+                reference(type.name)
+                text(" bean found in ")
+            }
+        }
     }
 
     data class Property(
         val kind: PropertyKind,
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) {
+            with(builder) {
+                text("$kind ")
+                reference(name)
+                text(" of ")
+            }
+        }
     }
 
     data class Project(
         val path: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) {
+            with(builder) {
+                text("project ")
+                reference(path)
+                text(" in ")
+            }
+        }
     }
 
     data class SystemProperty(
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) {
+            with(builder) {
+                text("system property ")
+                reference(name)
+                text(" set at ")
+            }
+        }
     }
 
-    abstract override fun toString(): String
-
     abstract override fun equals(other: Any?): Boolean
 
     abstract override fun hashCode(): Int
 
+    abstract override fun toString(): String
+
     /**
      * The shared logic for implementing `toString()` in subclasses.
+     *
+     * Renders this trace as a string (including a nested trace if it exists).
      */
     protected
-    fun asString(): String =
-        StringBuilder().apply {
-            sequence.forEach {
-                appendStringOf(it)
-            }
-        }.toString()
+    fun asString(): String = StructuredMessage.Builder().apply {
+        sequence.forEach {
+            it.describe(this)
```

Comment:
💅 Replace `apply` by `let` and name the builder parameter explicitly

Reply:
This missed the merge train. 

Was is that you couldn't easily tell what `this` referred to on line 287?

Reply:
`apply` is idiomatic when the receiver is used implicitly which is not the case here. By the way, I meant `also` not `let`.